cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,228 @@
+    List<Field<?>> primaryKey = getPrimaryKey(appId);
+    StructuredTable table = getNamespaceSourceControlMetadataTable();
+    Optional<StructuredRow> row = table.read(primaryKey);
+    if (row.isPresent()) {
+      StructuredRow nonNullRow = row.get();
+      String specificationHash = nonNullRow
+          .getString(StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD);
+      String commitId = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD);
+      Long lastSynced = nonNullRow.getLong(
+          StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD);
+      if (specificationHash == null && commitId == null && lastSynced == 0) {
+        return null;
+      }
+      return new SourceControlMeta(specificationHash, commitId, Instant.ofEpochMilli(lastSynced));
+    }
+    return null;
+  }
+
+  /**
+   * Sets the source control metadata for the specified application ID in the namespace.  Source
+   * control metadata will be null when the application is deployed in the namespace. It will be
+   * non-null when the application is pulled from the remote repository and deployed in the
+   * namespace.
+   *
+   * @param appId             {@link ApplicationId} for which the source control metadata is being
+   *                          set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void setNamespaceSourceControlMeta(ApplicationId appId,
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
+    StructuredTable scmTable = getNamespaceSourceControlMetadataTable();
+    scmTable.upsert(getNamespaceSourceControlMetaFields(appId, sourceControlMeta));
+  }
+
+  /**
+   * Updates the source control metadata of an application in a namespace. This function is called
+   * when application is pushed from the namespace to a remote repository. So,
+   * {@link SourceControlMeta} for this application should not be null and is updated in the
+   * {@code NamespaceSourceControlMetadata} table
+   *
+   * @param appId   {@link ApplicationId} whose source control metadata is being updated.
+   * @param scmMeta The source control metadata to update.
+   * @throws IOException                  if failed to write the metadata
+   * @throws ApplicationNotFoundException If the specified application ID is not found.
+   */
+  public void updateNamespaceSourceControlMeta(ApplicationId appId, SourceControlMeta scmMeta)
+      throws IOException, ApplicationNotFoundException {
+    StructuredTable namespaceSourceControlMetadataTable = getNamespaceSourceControlMetadataTable();
+    List<Field<?>> fields = getPrimaryKey(appId);
+    Optional<StructuredRow> existing = namespaceSourceControlMetadataTable.read(fields);
+    if (!existing.isPresent()) {
```

Comment:
we would like the source of truth of whether an application exists or not to be the applications table. Instead of having this as updateSourceControlMeta(), it would be better if this was just write(), with the responsibility of verifying that the app exists being left to the caller.

Reply:
Noted. Done